python/whylogs/api/writer/whylabs.py

```diff
@@ -688,65 +676,95 @@ def write(self, file: Writable, **kwargs: Any) -> Tuple[bool, str]:
             else:
                 view = _uncompound_dataset_profile(view, flags)
 
+        return view
+
+    def _upload_file(self, stamp, tmp_file, **kwargs) -> Tuple[bool, str]:
+        dataset_timestamp_epoch = int(stamp * 1000)
+        if self._transaction_id is not None:
+            # TODO: maybe check it's not a reference profile?
+            region = os.getenv("WHYLABS_UPLOAD_REGION", None)
+            client: TransactionsApi = self._get_or_create_transaction_client()
+            request = TransactionLogRequest(dataset_timestamp=dataset_timestamp_epoch, segment_tags=[], region=region)
+            result: AsyncLogResponse = client.log_transaction(self._transaction_id, request, **kwargs)
+            logger.info(f"Added profile {result.id} to transaction {self._transaction_id}")
+            response = self._do_upload(
+                dataset_timestamp=dataset_timestamp_epoch,
+                upload_url=result.upload_url,
+                profile_id=result.id,
+                profile_file=tmp_file,
+            )
+        else:
+            response = self._do_upload(
+                dataset_timestamp=dataset_timestamp_epoch,
+                profile_file=tmp_file,
+            )
+
+            return response
+
+    def _get_dataset_timestamp(self, view: DatasetProfileView) -> float:
+        utc_now = datetime.datetime.now(datetime.timezone.utc)
+        dataset_timestamp = view.dataset_timestamp or utc_now
+        stamp = dataset_timestamp.timestamp()
+        time_delta_seconds = utc_now.timestamp() - stamp
+
+        if time_delta_seconds < 0:
+            logger.warning(
+                f"About to upload a profile with a dataset_timestamp that is in the future: "
+                f"{time_delta_seconds}s old."
+            )
+        if time_delta_seconds > FIVE_YEARS_IN_SECONDS:
+            logger.error(
+                f"A profile being uploaded to WhyLabs has a dataset_timestamp of({dataset_timestamp}) "
+                f"compared to current datetime: {utc_now}. Uploads of profiles older than 5 years "
+                "might not be monitored in WhyLabs and may take up to 24 hours to show up."
+            )
+
+        if stamp <= 0:
+            logger.error(
+                f"Profiles should have timestamps greater than 0, but found a timestamp of {stamp}"
+                f" and current timestamp is {utc_now.timestamp()}, this is likely an error."
+            )
+
+        return stamp
+
+    @deprecated_alias(profile="file")
+    def write(self, file: Writable, zip_file: bool = False, **kwargs: Any) -> Tuple[bool, str]:
```

Comment:
We can't change this signature because it's part of the `Writer` interface. We can pull `zip_file` from `kwargs` here and then pass it along explicitly to the private methods.

Reply:
Actually, I wonder if it would be better to put the zip functionality in the `Writable` subclasses? I guess the `WhyLabsWriter` would still need to know if the `Writable` is zipped, but that could be arranged...